In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib
matplotlib.rcParams['figure.figsize'] = (20,10)

In [2]:
df = pd.read_csv('Bengaluru_House_Data.csv')
df

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [5]:
df.groupby('area_type')['area_type'].agg('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

In [6]:
df.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

I'm assuming that to predict prices 'area_type', 'availability', 'society', 'balcony' these columns are not important . so I'm going to drop them

In [7]:
df2 = df.drop(['area_type', 'availability', 'society', 'balcony'], axis=1)
df2

location       size total_sqft  bath   price
0      Electronic City Phase II      2 BHK       1056   2.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                      Kothanur      2 BHK       1200   2.0   51.00
...                         ...        ...        ...   ...     ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00
13316             Richards Town      4 BHK       3600   5.0  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00
13319              Doddathoguru      1 BHK        550   1.0   17.00

[13320 rows x 5 columns]

# Handling Null Values

In [8]:
df2.isna().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

My dataset have 13320 rows and the only 90 rows have null values, I can either fill the bath null avlues with their median or drop them. Since the difference between Non null rows and null rows is huge, I can simply and safely drop them.

In [10]:
df2.dropna(inplace=True)
df2.isna().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [11]:
df2

location       size total_sqft  bath   price
0      Electronic City Phase II      2 BHK       1056   2.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                      Kothanur      2 BHK       1200   2.0   51.00
...                         ...        ...        ...   ...     ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00
13316             Richards Town      4 BHK       3600   5.0  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00
13319              Doddathoguru      1 BHK        550   1.0   17.00

[13246 rows x 5 columns]

# Feature Engineering

Let's do some featurte engineering in 'size' column

In [12]:
df2['size'].value_counts()

size
2 BHK         5198
3 BHK         4286
4 Bedroom      818
4 BHK          577
3 Bedroom      546
1 BHK          531
2 Bedroom      329
5 Bedroom      296
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           57
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
14 BHK           1
13 BHK           1
12 Bedroom       1
27 BHK           1
43 Bedroom       1
16 BHK           1
19 BHK           1
18 Bedroom       1
Name: count, dtype: int64

In [14]:
df2['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

BHK stands for Bedrooms Hall Kitchen. And the number indicates how many bedrooms are there. Like 4 BHK means 4 bedrooms, 1 hall, 1 kitchen

In [15]:
df2['bedrooms'] = df2['size'].apply(lambda x: int(x.split(' ')[0]))
df2

location       size total_sqft  bath   price  bedrooms
0      Electronic City Phase II      2 BHK       1056   2.0   39.07         2
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00         4
2                   Uttarahalli      3 BHK       1440   2.0   62.00         3
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00         3
4                      Kothanur      2 BHK       1200   2.0   51.00         2
...                         ...        ...        ...   ...     ...       ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00         5
13316             Richards Town      4 BHK       3600   5.0  400.00         4
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00         2
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00         4
13319              Doddathoguru      1 BHK        550   1.0   17.00         1

[13246 rows x 6 columns]

Now let's take a look at the 'total_sqft' column

In [19]:
df2['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

At first glance we see the values are in string. and there are variations like range values

In [24]:
def is_float_possible(x):
    try:
        float(x)
    except:
        return False
    return True

In [27]:
df2[df2['total_sqft'].apply(is_float_possible)] # it's giving me all the rows that returns true for the function is applied

location       size total_sqft  bath   price  bedrooms
0      Electronic City Phase II      2 BHK       1056   2.0   39.07         2
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00         4
2                   Uttarahalli      3 BHK       1440   2.0   62.00         3
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00         3
4                      Kothanur      2 BHK       1200   2.0   51.00         2
...                         ...        ...        ...   ...     ...       ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00         5
13316             Richards Town      4 BHK       3600   5.0  400.00         4
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00         2
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00         4
13319              Doddathoguru      1 BHK        550   1.0   17.00         1

[13056 rows x 6 columns]

In [34]:
df2[~df2['total_sqft'].apply(is_float_possible)].head(10) # it's giving me all the rows that returns False for the function is applied

location       size      total_sqft  bath    price  bedrooms
30            Yelahanka      4 BHK     2100 - 2850   4.0  186.000         4
122              Hebbal      4 BHK     3067 - 8156   4.0  477.000         4
137  8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005         2
165            Sarjapur      2 BHK     1145 - 1340   2.0   43.490         2
188            KR Puram      2 BHK     1015 - 1540   2.0   56.800         2
410             Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500         1
549         Hennur Road      2 BHK     1195 - 1440   2.0   63.770         2
648             Arekere  9 Bedroom       4125Perch   9.0  265.000         9
661           Yelahanka      2 BHK     1120 - 1145   2.0   48.130         2
672        Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000         4

We can see other than range values there are also some values like Sq.metre, perch etc

In [68]:
def convert_into_float(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return (float(tokens[0]) + float(tokens[1]))/2
    else:
        try:
            return float(x)
        except:
            return x

In [69]:
df3 = df2.copy()

In [70]:
df3['total_sqft'] = df3['total_sqft'].apply(convert_into_float)
df3

location       size total_sqft  bath   price  bedrooms
0      Electronic City Phase II      2 BHK     1056.0   2.0   39.07         2
1              Chikka Tirupathi  4 Bedroom     2600.0   5.0  120.00         4
2                   Uttarahalli      3 BHK     1440.0   2.0   62.00         3
3            Lingadheeranahalli      3 BHK     1521.0   3.0   95.00         3
4                      Kothanur      2 BHK     1200.0   2.0   51.00         2
...                         ...        ...        ...   ...     ...       ...
13315                Whitefield  5 Bedroom     3453.0   4.0  231.00         5
13316             Richards Town      4 BHK     3600.0   5.0  400.00         4
13317     Raja Rajeshwari Nagar      2 BHK     1141.0   2.0   60.00         2
13318           Padmanabhanagar      4 BHK     4689.0   4.0  488.00         4
13319              Doddathoguru      1 BHK      550.0   1.0   17.00         1

[13246 rows x 6 columns]

In [78]:
df3[~df3['total_sqft'].apply(is_float_possible)].shape

(46, 6)

Other than range values there are more 46 different units values. we can just ignore them becase the number is so small caompare to 13246

In [79]:
df4 = df3[df3['total_sqft'].apply(is_float_possible)]
df4

location       size total_sqft  bath   price  bedrooms
0      Electronic City Phase II      2 BHK     1056.0   2.0   39.07         2
1              Chikka Tirupathi  4 Bedroom     2600.0   5.0  120.00         4
2                   Uttarahalli      3 BHK     1440.0   2.0   62.00         3
3            Lingadheeranahalli      3 BHK     1521.0   3.0   95.00         3
4                      Kothanur      2 BHK     1200.0   2.0   51.00         2
...                         ...        ...        ...   ...     ...       ...
13315                Whitefield  5 Bedroom     3453.0   4.0  231.00         5
13316             Richards Town      4 BHK     3600.0   5.0  400.00         4
13317     Raja Rajeshwari Nagar      2 BHK     1141.0   2.0   60.00         2
13318           Padmanabhanagar      4 BHK     4689.0   4.0  488.00         4
13319              Doddathoguru      1 BHK      550.0   1.0   17.00         1

[13200 rows x 6 columns]